### ISSUE TITLE
Regression in type inference with constructors with type arguments

### ISSUE THREAD

**Issue by adampauls** [microsoft/pyright#5369]
**Describe the bug**
The following code no longer typechecks as of 1.1.315:
```python
@dataclass(frozen=True, slots=True)
    class Test(Generic[G]):
        g: G

    x: Callable[[int], Test[int]] = Test
    assert x(5).g == 5
```

The error is
```
error: Expression of type "type[Test[G@Test]]" cannot be assigned to declared type "(int) -> Test[int]"
    Type "type[Test[G@Test]]" cannot be assigned to type "(int) -> Test[int]"
      Parameter 1: type "int" cannot be assigned to type "G@Test"
        Type "int" cannot be assigned to type "G@Test" (reportGeneralTypeIssues)
```

The code runs fine. The following code typechecks: 

```python
@dataclass(frozen=True, slots=True)
    class Test(Generic[G]):
        g: G

    x: Callable[[int], Test[int]] = Test[int] # note the type argument here
    assert x(5).g == 5
```

but fails at runtime with 
```
/Library/Frameworks/Python.framework/Versions/3.11/lib/python3.11/typing.py:1250: in __call__
    result.__orig_class__ = self
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

self = test_xxx.<locals>.Test(g=5), name = '__orig_class__'
value = tests.anchor.test_events.test_xxx.<locals>.Test[int]

>   ???
E   TypeError: super(type, obj): obj must be an instance or subtype of type
```

**Expected behavior**
I believe the old pyright behavior was correct given the runtime behavior. 

**VS Code extension or command-line**
Command-line version 1.1.315.

**Comment by adampauls** [comment]
I was too hasty in following the bug. This is a minimization of my actual problem, which involves a decorator that expects a factory method to which I had been passing class names:

```python
def publishes_events(
    event_factory: Callable[
        Concatenate[MethodCallEventType, Publisher, Output | None, Exception | None, Params],
        MethodCallEventT,
    ],
) -> SyncOrAsyncMethod[Publisher, Params, Output]:
   ...

class SyncOrAsyncMethod(Protocol[Slf, Params, Return]):
    @overload
    def __call__(self, func: Callable[Concatenate[Slf, Params], Return]) -> Callable[Concatenate[Slf, Params], Return]:
        ...

    @overload
    def __call__(
        self,
        func: Callable[Concatenate[Slf, Params], Awaitable[Return]],
    ) -> Callable[Concatenate[Slf, Params], Awaitable[Return]]:
        ...
``` 

As of 1.1.313, passing a bare classname (like `Test`) for a class that takes type parameters was working fine, but as of 1.1.315 it is failing. If I add the type arguments, then Pyright is happy but the decorator crashes at runtime with the same error as above.

However, it appears that the code I have above never type-checked, at least not for 1.1.312 <= <= 1.1.315. Not sure what's going on here.

**Comment by erictraut** [comment]
Could you please provide a minimal, self-contained code sample? The sample above references a bunch of types that are not defined within the sample. I could guess at how they are defined, but it would be better if you could provide them. I presume that some of the identifiers refer to type variables, others refer to classes.

**Comment by erictraut** [comment]
As for your top example, I can't explain why the runtime is generating an exception here. This strikes me as a runtime bug. If I remove the `frozen=True` or the `slots=True`, it works fine. Here's a simplified example if you care to report it as a bug against the Python runtime.

```python
from dataclasses import dataclass
from typing import Generic, TypeVar

T = TypeVar("T")
@dataclass(frozen=True, slots=True)
class Test(Generic[T]):
    x: T

Test[int](3)
```

Here's the output:
```
Traceback (most recent call last):
  File "<redacted.py>", line 12, in <module>
    Test[int](3)
  File "/Library/Frameworks/Python.framework/Versions/3.11/lib/python3.11/typing.py", line 1253, in __call__
    result.__orig_class__ = self
    ^^^^^^^^^^^^^^^^^^^^^
  File "<string>", line 5, in __setattr__
TypeError: super(type, obj): obj must be an instance or subtype of type
```

**Comment by adampauls** [comment]
I will try (again) to minimize. Do you believe the very first (self-contained) example should fail to typecheck, even though the type can be inferred?

**Comment by erictraut** [comment]
Yes, I agree that the top example should type check. It does with other generic functions, like this:

```python
def func(x: G) -> Test[G]: ...
y: Callable[[int], Test[int]] = func
```

So it should work consistently when applied to a constructor call. I'll investigate further. 

Is your second sample the same problem? If so, then it's unnecessary for you to provide an updated version. If you think it's different from the first problem, then I'd appreciate a complete sample so I can repro it.

**Comment by mgmira** [comment]
For the runtime error, I've encountered a similar error with python 3.11 a few months ago, and there is a related issue in the bug tracker https://github.com/python/cpython/issues/90562

**Comment by erictraut** [comment]
This is addressed in pyright 1.1.316, which I just published. It will also be included in this week's insiders build of pylance.
